unicef/giga-data-ingestion
# [US41-02] QoS school list model

## What type of PR is this?


- `feat`: Commits that add a new feature

## Summary

What does this PR do

- Adds a `downgrade` task
- Adds a model based on the [schema](https://docs.google.com/document/d/1t30K13iMyPyfvHBqlj7Iy4ASaYsU3CcHpGyiihU_6ks/edit#heading=h.2pkld575kj4y) needed by the DE QOS school-list pipeline


DE expects the following table
```
CREATE TABLE IF NOT EXISTS `${schema_name}`.`${table_name}` (
  request_method STRING NOT NULL, # ENUM GET/POST
  api_endpoint STRING NOT NULL,
  data_key STRING,
  school_id_key STRING NOT NULL,

  query_parameters STRING, # will need to rely on frontend to validate that keys and values exist
 request_body STRING, # will need to rely on frontend to validate that keys and values exist
  authorization_type STRING, # ENUM BEARER_TOKEN/BASIC_AUTH/API_KEY/NULL
  bearer_auth_bearer_token STRING, # NOT NULL if authorization_type == BEARER_TOKEN
  basic_auth_username STRING, # NOT NULL if authorization_type == BASIC_AUTH
  basic_auth_password STRING, # NOT NULL if authorization_type == BASIC_AUTH
  api_auth_api_key STRING, # NOT NULL if authorization_type == API_KEY
  api_auth_api_value STRING, # NOT NULL if authorization_type == API_KEY

  pagination_type STRING, # ENUM PAGE_NUMBER/LIMIT_OFFSET/NULL # page number/limit offset/ null
  size INT, # NOT NULL if paginated == PAGE_NUMBER/LIMIT_OFFSET #records per page
  page_size_key STRING, # NOT NULL if paginated == PAGE_NUMBER/LIMIT_OFFSET
  send_query_in STRING, # ENUM BODY/QUERY_PARAMETERS/HEADERS, # NOT NULL if paginated == PAGE_NUMBER/LIMIT_OFFSET # pass the info via query / body / headers // what info tho

  page_number_key STRING, # NOT NULL if paginated == PAGE_NUMBER # this is their naming convention of what current_page_selected is
  page_starts_with INT, # NOT NULL if paginated == PAGE_NUMBER # if 0 or 1 indexed

  page_offset_key STRING, # NOT NULL if paginated == LIMIT_OFFSET
 enabled BOOL

 date_created TIMESTAMP
 date_modified TIMESTAMP
 user_id STRING
 user_email STRING
 

)
# ONE TABLE FOR SCHOOL LIST, ONE TABLE FOR ACTUAL QOS DATA?

```

## How to test

1. `Task migrate`
2. Check if fields and enums appear in database
3. Optionally check if you can downgrade the database to an earlier revision , in this case`d0ecb42fa08e`

https://github.com/unicef/giga-data-ingestion/assets/122899250/7272fff7-3a0a-400a-a743-63383d61b4b9



## Link to Jira/Asana/Airtable task (if applicable)

[link](https://app.asana.com/0/1206284549613121/1206643852697936)
